crates/astria-conductor/src/data_availability.rs

```diff
@@ -0,0 +1,382 @@
+use std::time::Duration;
+
+use astria_sequencer_relayer::data_availability::{
+    CelestiaClient,
+    SequencerNamespaceData,
+    SignedNamespaceData,
+};
+use astria_sequencer_types::Namespace;
+use color_eyre::eyre::{
+    self,
+    WrapErr as _,
+};
+use tokio::{
+    select,
+    sync::{
+        mpsc,
+        oneshot,
+    },
+    task::{
+        self,
+        JoinError,
+        JoinHandle,
+        JoinSet,
+    },
+};
+use tokio_util::task::JoinMap;
+use tracing::{
+    error,
+    info,
+    instrument,
+    warn,
+    Instrument,
+};
+
+use crate::{
+    block_verifier::{
+        self,
+        BlockVerifier,
+    },
+    // config::Config,
+    executor,
+    types::SequencerBlockSubset,
+};
+
+pub(crate) struct Reader {
+    /// The channel used to send messages to the executor task.
+    executor_tx: executor::Sender,
+
+    /// The client used to communicate with Celestia.
+    celestia_client: CelestiaClient,
+
+    /// The between the reader waits until it queries celestia for new blocks
+    celestia_poll_interval: Duration,
+
+    /// the last block height fetched from Celestia
+    current_block_height: u64,
+
+    block_verifier: BlockVerifier,
+
+    /// Namespace ID
+    namespace: Namespace,
+
+    get_latest_height: Option<JoinHandle<eyre::Result<u64>>>,
+
+    /// A map of in-flight queries to celestia for new sequencer blocks
+    get_sequencer_datas:
+        JoinMap<u64, eyre::Result<Vec<SignedNamespaceData<SequencerNamespaceData>>>>,
+
+    process_sequencer_datas: JoinMap<u64, eyre::Result<Vec<SequencerBlockSubset>>>,
+
+    shutdown: oneshot::Receiver<()>,
+}
+
+impl Reader {
+    /// Creates a new Reader instance and returns a command sender.
+    pub(crate) async fn new(
+        celestia_node_url: &str,
+        celestia_bearer_token: &str,
+        celestia_poll_interval: Duration,
+        executor_tx: executor::Sender,
+        block_verifier: BlockVerifier,
+        namespace: Namespace,
+        shutdown: oneshot::Receiver<()>,
+    ) -> eyre::Result<Self> {
+        let celestia_client = CelestiaClient::builder()
+            .endpoint(celestia_node_url)
+            .bearer_token(celestia_bearer_token)
+            .build()
+            .wrap_err("failed creating celestia client")?;
+
+        // TODO: we should probably pass in the height we want to start at from some genesis/config
+        // file
+        let current_block_height = celestia_client.get_latest_height().await?;
+        info!(da_height = current_block_height, "creating Reader");
+
+        Ok(Self {
+            executor_tx,
+            celestia_client,
+            celestia_poll_interval,
+            current_block_height,
+            get_latest_height: None,
+            get_sequencer_datas: JoinMap::new(),
+            process_sequencer_datas: JoinMap::new(),
```

Comment:
nit: this is data_availability data and we need to process it to make sure it comes from the sequencer, the output of processing is sequencer data

Reply:
The naming comes from `SequencerNamespaceData`, so the thinking is that we are processing multiple `sequencer_(namespace_)datas`. The naming is very unfortunate.

Reply:
Renamed it to `verify_sequencer_blobs_and_assemble_rollups` (as well as its joinmap).